### PR TITLE
add bandwidth_max to config, issue #492

### PR DESCRIPTION
**Pull request by gijzelaerr**
So this PR got much bigger than I was hoping, but basically:

* adds bandwidth_max to config, if set will limit bandwidth for image band association
* Removes most of the SQL functions and replaced with python ones
* Cut out a lot of the ORM usage
* Fix PEP8 problems that I came by
* Add frequency bands per dataset, not global per DB

the ORM removal is not finished yet, but I've most of the methods to `tkp.db.general`  so it becomes easier to replace the Image, Dataset and Extractedsources objects with the SQLAlchemy equivalents. Fun fact, the ORM is almost only used in the test suite.

### CODE REVIEW

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
retest this please

**Comment by gijzelaerr** [comment]
I'm still having some issues configuring jenkins. but the PR build is actually GREEN now:

https://jenkins.transientskp.org/job/tkp_pullrequest/348/

**Comment by gijzelaerr** [comment]
retest this please

File: CHANGES.md [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+
+the band determination logic has changed. Before all bands where split
+into 1 MHz intervals and associated as such. With this release  images are
+put in the same band if their bandwidths overlap.
+
+We added an option to limit the bandwidth used for band assoication ([#492][]).
+Limiting the bandwidth for an image is done by setting `bandwidth_max`
+in *job_params.cfg* under the `persistence section`. E.g.:
+
+    [persistense]
```
Comment (timstaley) [inline CHANGES.md]: typo

File: documentation/userref/config/job_params_cfg.rst [hunk 2]
```diff
@@ -40,6 +40,11 @@ The parameters in this file are defined as follows:
    Integer. Determines the size of the subsection used for RMS measurement:
    the central ``1/f`` of the image will be used (where f=rms_est_fraction).
 
+``bandwidth``
```
Comment (timstaley) [inline documentation/userref/config/job_params_cfg.rst]: --> 'bandwidth_max'
I would maybe tweak the docstring a bit too:
"Limit the maximum bandwidth used when determining if two images belong to the same 'band' grouping."

File: tkp/db/alchemy/image.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                 rms_min=None, rms_max=None, detection_thresh=None, analysis_thresh=None):
+    """
+    Insert an image for a given dataset.
+
+    Args:
+        session (sqlalchemy.orm.session.Session): A SQLalchemy sessions
+        dataset (int): ID of parent dataset.
+        freq_eff: See :ref:`Image table definitions <schema-image>`.
+        freq_bw: See :ref:`Image table definitions <schema-image>`.
+        freq_bw_max (float): Optional override for freq_bw, not used if 0.
```
Comment (timstaley) [inline tkp/db/alchemy/image.py]: What was the reasoning behind adding a freq_bw_max field to the image? Why not just use whatever's currently in the job_config?

File: tkp/db/alchemy/image.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        high = str(freq_eff + bw_half)
+        filter_ = " between freq_low and freq_high"
+    else:
+        bw_half = freq_bw_max / 2
+        low = str(freq_eff - bw_half)
+        high = str(freq_eff + bw_half)
+        filter_ = " between freq_central - %s and freq_central + %s" % (bw_half, bw_half)
+
+    band = session.query(Frequencyband). \
+        filter((Frequencyband.dataset == dataset) & (text(low + filter_) | text(high + filter_))).first()
```
Comment (timstaley) [inline tkp/db/alchemy/image.py]: Hmm, I think I see a failure / edge case, if I've understood this correctly. To me it looks like this code is calculating the limits of the current image bandwidth, and then seeing if those limits lie within previously defined bands?

But: what if the new image has identical central frequency, but slightly larger bandwidth than the previous image (by like, 1% at low and high end). They should probably go in the same band right? But the low end won't sit within the previous ranges, and neither will the high end. 

Maybe we need to actually work with the current image central bandwidth, and I don't know, calculate distances between band centres, or something?

File: tkp/db/alchemy/image.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        centre_decl (float): center DECL coordinate
+        xtr_radius (float): The extraction radius
+
+     returns:
+        tkp.db.models.Skyregion: a SQLalchemy skyregion
+     """
+    skyregion = session.query(Skyregion).filter(Skyregion.dataset == dataset,
+                                                Skyregion.centre_ra == centre_ra,
+                                                Skyregion.centre_decl == centre_decl,
+                                                Skyregion.xtr_radius == xtr_radius).first()
```
Comment (timstaley) [inline tkp/db/alchemy/image.py]: Using `.first` works fine here, but I would say [`.one_or_none`](http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.one_or_none) gives better documentation of intent - we don't expect there to ever be multiple identical Skyregions, so `first` looks a bit weird to me.

Comment (gijzelaerr) [inline tkp/db/alchemy/image.py]: ah, i didn't knew that one existed also. You are getting more experienced with SQLAlchemy than me ;)

File: tkp/db/alchemy/image.py (hunk 4 above)
Comment (gijzelaerr) [inline tkp/db/alchemy/image.py]: or we test on `max(low, freq_low) <= min(high, freq_high)`, that always returns the overlapping bands. what do you think?

Comment (timstaley) [inline tkp/db/alchemy/image.py]: I'm having trouble engaging brain on this, so how about I just refer you to http://stackoverflow.com/questions/3269434

Comment (gijzelaerr) [inline tkp/db/alchemy/image.py]: heh yes just updated my comment

**Comment by timstaley** [comment]
Great, looks good to me.

**Comment by gijzelaerr** [comment]
lets get crazy and just do it

**Pull request merged**